File: python/csf-analysis/src/csf_analysis/stemmer_en.py
import re
from dataclasses import dataclass
from typing import Callable

_VOWEL = re.compile(r"[aeiou]")


def _is_consonant(word: str, i: int) -> bool:
    if i < 0 or i >= len(word):
        return False
    c = word[i]
    if _VOWEL.match(c):
        return False
    if c != "y":
        return True
    return i == 0 or not _is_consonant(word, i - 1)


def _measure(word: str) -> int:
    i = 0
    n = len(word)
    while i < n and _is_consonant(word, i):
        i += 1
    m = 0
    while i < n:
        while i < n and not _is_consonant(word, i):
            i += 1
        if i >= n:
            break
        while i < n and _is_consonant(word, i):
            i += 1
        m += 1
    return m


def _has_vowel(word: str) -> bool:
    return any(not _is_consonant(word, i) for i in range(len(word)))


def _ends_double_consonant(word: str) -> bool:
    n = len(word)
    if n < 2:
        return False
    return (
        word[n - 1] == word[n - 2]
        and _is_consonant(word, n - 1)
        and _is_consonant(word, n - 2)
    )


def _ends_cvc(word: str) -> bool:
    n = len(word)
    if n < 3:
        return False
    last = word[n - 1]
    return (
        _is_consonant(word, n - 3)
        and not _is_consonant(word, n - 2)
        and _is_consonant(word, n - 1)
        and last not in ("w", "x", "y")
    )


@dataclass(frozen=True)
class _Rule:
    suffix: str
    replacement: str
    condition: Callable[[str], bool] | None = None


def _apply_rules(word: str, rules: list[_Rule]) -> str:
    for rule in rules:
        if not word.endswith(rule.suffix):
            continue
        stem = word[: len(word) - len(rule.suffix)]
        if rule.condition and not rule.condition(stem):
            return word
        return stem + rule.replacement
    return word


def _step1a(word: str) -> str:
    return _apply_rules(
        word,
        [
            _Rule("sses", "ss"),
            _Rule("ies", "i"),
            _Rule("ss", "ss"),
            _Rule("s", ""),
        ],
    )


def _restore_after_step1b(stem: str) -> str:
    if stem.endswith("at") or stem.endswith("bl") or stem.endswith("iz"):
        return stem + "e"
    if _ends_double_consonant(stem) and stem[-1] not in ("l", "s", "z"):
        return stem[:-1]
    if _measure(stem) == 1 and _ends_cvc(stem):
        return stem + "e"
    return stem


def _step1b(word: str) -> str:
    if word.endswith("eed"):
        stem = word[:-3]
        return stem + "ee" if _measure(stem) > 0 else word
    for suffix in ("ed", "ing"):
        if not word.endswith(suffix):
            continue
        stem = word[: len(word) - len(suffix)]
        if not _has_vowel(stem):
            continue
        return _restore_after_step1b(stem)
    return word


def _step1c(word: str) -> str:
    if not word.endswith("y"):
        return word
    stem = word[:-1]
    if len(stem) == 0 or not _has_vowel(stem):
        return word
    return stem + "i"


def _m0(stem: str) -> bool:
    return _measure(stem) > 0


def _step2(word: str) -> str:
    return _apply_rules(
        word,
        [
            _Rule("ational", "ate", _m0),
            _Rule("tional", "tion", _m0),
            _Rule("enci", "ence", _m0),
            _Rule("anci", "ance", _m0),
            _Rule("izer", "ize", _m0),
            _Rule("bli", "ble", _m0),
            _Rule("alli", "al", _m0),
            _Rule("entli", "ent", _m0),
            _Rule("eli", "e", _m0),
            _Rule("ousli", "ous", _m0),
            _Rule("ization", "ize", _m0),
            _Rule("ation", "ate", _m0),
            _Rule("ator", "ate", _m0),
            _Rule("alism", "al", _m0),
            _Rule("iveness", "ive", _m0),
            _Rule("fulness", "ful", _m0),
            _Rule("ousness", "ous", _m0),
            _Rule("aliti", "al", _m0),
            _Rule("iviti", "ive", _m0),
            _Rule("biliti", "ble", _m0),
            _Rule("logi", "log", _m0),
        ],
    )


def _step3(word: str) -> str:
    return _apply_rules(
        word,
        [
            _Rule("icate", "ic", _m0),
            _Rule("ative", "", _m0),
            _Rule("alize", "al", _m0),
            _Rule("iciti", "ic", _m0),
            _Rule("ical", "ic", _m0),
            _Rule("ful", "", _m0),
            _Rule("ness", "", _m0),
        ],
    )


def _m1(stem: str) -> bool:
    return _measure(stem) > 1


def _step4(word: str) -> str:
    return _apply_rules(
        word,
        [
            _Rule("al", "", _m1),
            _Rule("ance", "", _m1),
            _Rule("ence", "", _m1),
            _Rule("er", "", _m1),
            _Rule("ic", "", _m1),
            _Rule("able", "", _m1),
            _Rule("ible", "", _m1),
            _Rule("ant", "", _m1),
            _Rule("ement", "", _m1),
            _Rule("ment", "", _m1),
            _Rule("ent", "", _m1),
            _Rule(
                "ion",
                "",
                lambda stem: _m1(stem) and (stem.endswith("s") or stem.endswith("t")),
            ),
            _Rule("ou", "", _m1),
            _Rule("ism", "", _m1),
            _Rule("ate", "", _m1),
            _Rule("iti", "", _m1),
            _Rule("ous", "", _m1),
            _Rule("ive", "", _m1),
            _Rule("ize", "", _m1),
        ],
    )


def _step5a(word: str) -> str:
    if not word.endswith("e"):
        return word
    stem = word[:-1]
    m = _measure(stem)
    if m > 1 or (m == 1 and not _ends_cvc(stem)):
        return stem
    return word


def _step5b(word: str) -> str:
    if word.endswith("ll") and _measure(word[:-1]) > 1:
        return word[:-1]
    return word


_ASCII_LETTERS_ONLY = re.compile(r"^[a-z]+$")


def stem_english(word: str) -> str:
    if len(word) <= 2 or not _ASCII_LETTERS_ONLY.match(word):
        return word
    result = _step1a(word)
    result = _step1b(result)
    result = _step1c(result)
    result = _step2(result)
    result = _step3(result)
    result = _step4(result)
    result = _step5a(result)
    result = _step5b(result)
    return result

File: packages/analysis/src/stemmer-en.ts
/**
 * The classic Porter stemming algorithm (M.F. Porter, "An algorithm for
 * suffix stripping", 1980) for English -- the affix-stripping algorithm
 * docs/guides/internationalization.md#stemming commits to for languages with
 * "a good affix-stripping stemmer". Deliberately the original 1980
 * algorithm, not the later Snowball-framework "Porter2" English
 * stemmer (a distinct, incompatible rule set) -- the original is what
 * this module implements and what its test suite verifies against a
 * public 23,531-word reference vocabulary, and upgrading to Porter2 is
 * a separate, from-scratch rewrite rather than an incremental tweak.
 */

const VOWEL = /[aeiou]/;

/** A letter at `i` is a consonant unless it's a plain vowel, or `y` immediately after a vowel. */
function isConsonant(word: string, i: number): boolean {
  const c = word[i];
  if (c === undefined) return false;
  if (VOWEL.test(c)) return false;
  if (c !== "y") return true;
  return i === 0 || !isConsonant(word, i - 1);
}

/**
 * The word's consonant/vowel pattern collapses to [C](VC)^m[V]; `m` is
 * the number of VC repetitions, computed by scanning past the leading
 * consonant run (if any) and counting each vowel-run-to-consonant-run
 * transition.
 */
function measure(word: string): number {
  let i = 0;
  const n = word.length;
  while (i < n && isConsonant(word, i)) i++;
  let m = 0;
  while (i < n) {
    while (i < n && !isConsonant(word, i)) i++;
    if (i >= n) break;
    while (i < n && isConsonant(word, i)) i++;
    m++;
  }
  return m;
}

/** Whether `word` contains a vowel anywhere (the "*v*" condition). */
function hasVowel(word: string): boolean {
  for (let i = 0; i < word.length; i++) {
    if (!isConsonant(word, i)) return true;
  }
  return false;
}

/** Whether `word` ends in two identical consonants (the "*d" condition). */
function endsDoubleConsonant(word: string): boolean {
  const n = word.length;
  if (n < 2) return false;
  return (
    word[n - 1] === word[n - 2] &&
    isConsonant(word, n - 1) &&
    isConsonant(word, n - 2)
  );
}

/**
 * Whether `word` ends consonant-vowel-consonant with the final
 * consonant not w, x, or y (the "*o" condition) -- used to decide
 * whether a short CVC stem like "hop"/"trap" needs a trailing `e`
 * restored after suffix removal.
 */
function endsCvc(word: string): boolean {
  const n = word.length;
  if (n < 3) return false;
  const last = word[n - 1] as string;
  return (
    isConsonant(word, n - 3) &&
    !isConsonant(word, n - 2) &&
    isConsonant(word, n - 1) &&
    last !== "w" &&
    last !== "x" &&
    last !== "y"
  );
}

interface Rule {
  suffix: string;
  replacement: string;
  /** Extra condition on the stem (the part before `suffix`), beyond the suffix match itself. */
  condition?: (stem: string) => boolean;
}

/**
 * Applies the rule for the *longest* suffix (in list order, so `rules`
 * must list any mutually-nesting suffixes -- e.g. "ement"/"ment"/"ent"
 * -- longest first) that textually matches `word`. Once that one rule
 * is found, its condition is checked exactly once: if it fails, `word`
 * is returned unchanged rather than falling through to try a shorter
 * nested suffix (e.g. a word ending "...ement" whose "ement" rule's
 * condition fails must not then be reinterpreted as ending in the
 * shorter "ent", even though it textually also matches).
 */
function applyRules(word: string, rules: Rule[]): string {
  for (const rule of rules) {
    if (!word.endsWith(rule.suffix)) continue;
    const stem = word.slice(0, word.length - rule.suffix.length);
    if (rule.condition && !rule.condition(stem)) return word;
    return stem + rule.replacement;
  }
  return word;
}

function step1a(word: string): string {
  return applyRules(word, [
    { suffix: "sses", replacement: "ss" },
    { suffix: "ies", replacement: "i" },
    { suffix: "ss", replacement: "ss" },
    { suffix: "s", replacement: "" },
  ]);
}

/** Post-processing shared by the `ed`/`ing` rules of step 1b once one of them fires. */
function restoreAfterStep1b(stem: string): string {
  if (stem.endsWith("at") || stem.endsWith("bl") || stem.endsWith("iz")) {
    return `${stem}e`;
  }
  if (endsDoubleConsonant(stem) && !/[lsz]$/.test(stem)) {
    return stem.slice(0, -1);
  }
  if (measure(stem) === 1 && endsCvc(stem)) {
    return `${stem}e`;
  }
  return stem;
}

function step1b(word: string): string {
  if (word.endsWith("eed")) {
    const stem = word.slice(0, -3);
    return measure(stem) > 0 ? `${stem}ee` : word;
  }
  for (const suffix of ["ed", "ing"]) {
    if (!word.endsWith(suffix)) continue;
    const stem = word.slice(0, word.length - suffix.length);
    if (!hasVowel(stem)) continue;
    return restoreAfterStep1b(stem);
  }
  return word;
}

function step1c(word: string): string {
  if (!word.endsWith("y")) return word;
  const stem = word.slice(0, -1);
  if (stem.length === 0 || !hasVowel(stem)) return word;
  return `${stem}i`;
}

const m0 = (stem: string) => measure(stem) > 0;

function step2(word: string): string {
  return applyRules(word, [
    { suffix: "ational", replacement: "ate", condition: m0 },
    { suffix: "tional", replacement: "tion", condition: m0 },
    { suffix: "enci", replacement: "ence", condition: m0 },
    { suffix: "anci", replacement: "ance", condition: m0 },
    { suffix: "izer", replacement: "ize", condition: m0 },
    { suffix: "bli", replacement: "ble", condition: m0 },
    { suffix: "alli", replacement: "al", condition: m0 },
    { suffix: "entli", replacement: "ent", condition: m0 },
    { suffix: "eli", replacement: "e", condition: m0 },
    { suffix: "ousli", replacement: "ous", condition: m0 },
    { suffix: "ization", replacement: "ize", condition: m0 },
    { suffix: "ation", replacement: "ate", condition: m0 },
    { suffix: "ator", replacement: "ate", condition: m0 },
    { suffix: "alism", replacement: "al", condition: m0 },
    { suffix: "iveness", replacement: "ive", condition: m0 },
    { suffix: "fulness", replacement: "ful", condition: m0 },
    { suffix: "ousness", replacement: "ous", condition: m0 },
    { suffix: "aliti", replacement: "al", condition: m0 },
    { suffix: "iviti", replacement: "ive", condition: m0 },
    { suffix: "biliti", replacement: "ble", condition: m0 },
    { suffix: "logi", replacement: "log", condition: m0 },
  ]);
}

function step3(word: string): string {
  return applyRules(word, [
    { suffix: "icate", replacement: "ic", condition: m0 },
    { suffix: "ative", replacement: "", condition: m0 },
    { suffix: "alize", replacement: "al", condition: m0 },
    { suffix: "iciti", replacement: "ic", condition: m0 },
    { suffix: "ical", replacement: "ic", condition: m0 },
    { suffix: "ful", replacement: "", condition: m0 },
    { suffix: "ness", replacement: "", condition: m0 },
  ]);
}

const m1 = (stem: string) => measure(stem) > 1;

function step4(word: string): string {
  return applyRules(word, [
    { suffix: "al", replacement: "", condition: m1 },
    { suffix: "ance", replacement: "", condition: m1 },
    { suffix: "ence", replacement: "", condition: m1 },
    { suffix: "er", replacement: "", condition: m1 },
    { suffix: "ic", replacement: "", condition: m1 },
    { suffix: "able", replacement: "", condition: m1 },
    { suffix: "ible", replacement: "", condition: m1 },
    { suffix: "ant", replacement: "", condition: m1 },
    { suffix: "ement", replacement: "", condition: m1 },
    { suffix: "ment", replacement: "", condition: m1 },
    { suffix: "ent", replacement: "", condition: m1 },
    {
      suffix: "ion",
      replacement: "",
      condition: (stem) =>
        m1(stem) && (stem.endsWith("s") || stem.endsWith("t")),
    },
    { suffix: "ou", replacement: "", condition: m1 },
    { suffix: "ism", replacement: "", condition: m1 },
    { suffix: "ate", replacement: "", condition: m1 },
    { suffix: "iti", replacement: "", condition: m1 },
    { suffix: "ous", replacement: "", condition: m1 },
    { suffix: "ive", replacement: "", condition: m1 },
    { suffix: "ize", replacement: "", condition: m1 },
  ]);
}

function step5a(word: string): string {
  if (!word.endsWith("e")) return word;
  const stem = word.slice(0, -1);
  const m = measure(stem);
  if (m > 1 || (m === 1 && !endsCvc(stem))) return stem;
  return word;
}

function step5b(word: string): string {
  if (word.endsWith("ll") && measure(word.slice(0, -1)) > 1) {
    return word.slice(0, -1);
  }
  return word;
}

/** Every rule above assumes plain a-z letters; anything else (diacritics, digits, other scripts) is returned unchanged rather than stemmed nonsensically. */
const ASCII_LETTERS_ONLY = /^[a-z]+$/;

/**
 * Stems an already-lowercased English word via the classic Porter
 * algorithm's five steps, applied in sequence. Words too short to
 * carry an inflectional suffix (len <= 2), or containing anything
 * outside plain a-z (an English profile doesn't fold diacritics, so an
 * accented loanword like "café" can reach here unmodified), are
 * returned unchanged, same as the reference algorithm's ASCII-only
 * scope.
 */
export function stemEnglish(word: string): string {
  if (word.length <= 2 || !ASCII_LETTERS_ONLY.test(word)) return word;
  let result = step1a(word);
  result = step1b(result);
  result = step1c(result);
  result = step2(result);
  result = step3(result);
  result = step4(result);
  result = step5a(result);
  result = step5b(result);
  return result;
}
